src/dtree.ml

```diff
@@ -0,0 +1,648 @@
+  val of_list : 'a list -> 'a t
+
+  (** [length v] is the number of elements in [v]. *)
+  val length : 'a t -> int
+
+  (** Prefix and suffix of the substrate. *)
+  type 'a prefix
+  type 'a suffix
+
+  (** [destruct v i] returns a triplet [(r, e, o)] with [r] being the
+      elements from 0 to [i - 1], [e] the [i]th element and [o] the elements
+      from [i + 1] to the end of [v]. *)
+  val destruct : 'a t -> int -> 'a prefix * 'a * 'a suffix
+
+  (** [restruct r n o] is the concatenation of three stacks [r] [n] and
+      [o]. *)
+  val restruct : 'a prefix -> 'a list -> 'a suffix -> 'a t
+end
+
+module ReductionStack : reduction_substrate =
+struct
+  type 'a t = 'a list
+  type 'a prefix = 'a list
+  type 'a suffix = 'a list
+  let empty = []
+  let is_empty l = (=) [] l
+  let of_list l = l
+  let length = List.length
+  let destruct e i =
+    if i < 0 then invalid_arg "RedStack.destruct";
+    let rec destruct l i r =
+      match (r, i) with
+      | ([]  , _) -> raise Not_found
+      | (v::r, 0) -> (l, v, r)
+      | (v::r, i) -> destruct (v :: l) (i - 1) r
+    in
+    destruct [] i e
+  let restruct l c r =
+    let rec insert acc l =
+      match l with
+      | []   -> acc
+      | x::l -> insert (x :: acc) l
+    in
+    insert (c @ r) l
+end
+
+(** {3 Operators on trees} *)
+
+(** [iter l n f t] is a generic iterator on trees; with function [l] performed
+    on leaves, function [n] performed on nodes, [f] returned in case of
+    {!constructor:Fail} on tree [t]. *)
+let iter : do_leaf:(int IntMap.t -> action -> 'a) ->
+  do_node:(int -> bool -> 'a ConsMap.t -> 'a option -> 'a) ->
+  do_fetch:(bool -> 'a -> 'a) ->
+  fail:'a -> t -> 'a = fun ~do_leaf ~do_node ~do_fetch ~fail t ->
+  let rec loop = function
+    | Leaf(pa, a)                                 -> do_leaf pa a
+    | Fail                                        -> fail
+    | Node({ swap ; store ; children ; default }) ->
+       do_node swap store
+         (ConsMap.map (fun c -> loop c) children)
+         (Option.map loop default)
+    | Fetch(store, next)                          -> do_fetch store (loop next)
+  in
+  loop t
+
+(** [capacity t] computes the capacity of tree [t].  During evaluation, some
+    terms that are being filtered by the patterns need to be saved in order to
+    be bound in the right hand side of the rule.  The capacity is an upper
+    bound of the number of terms to be saved. *)
+let capacity : t -> int =
+  let do_leaf _ _ = 0 in
+  let fail = 0 in
+  let do_node _ st ch de =
+    let _, chdepths = List.split (ConsMap.bindings ch) in
+    let dedepth = Option.get de 0 in
+    List.extremum (>) (dedepth :: chdepths) +
+      (if st then 1 else 0) in
+  let do_fetch st ne = (if st then 1 else 0) + ne in
+  iter ~do_leaf:do_leaf ~fail:fail ~do_node:do_node ~do_fetch:do_fetch
+
+(** {3 Constructors} *)
+
+let cons_of_term : term -> constructor = fun te ->
+  let hs, args = Basics.get_args te in
+  let arity = List.length args in
```

Comment:
Remark: get_args could be improved by returning the length of args too.

Reply:
Indeed, but I'd have to edit a lot of calls in surrounding code, so perhaps it shouldn't be done in this PR.